plugins/Phaser/PhaserEffect.cpp

```diff
@@ -0,0 +1,623 @@
+
+			Note that the DC offset added here by this distortion is
+			removed later on.
+			*/
+			const float trueDistVal = distortion * 1.5;
+			tempAdd[0] = abs(tempAdd[0] + 1.5 - trueDistVal) - 1.5 + trueDistVal;
+			tempAdd[1] = abs(tempAdd[1] + 1.5 - trueDistVal) - 1.5 + trueDistVal;
+		}
+
+		// Remove DC Offset
+		for (int i = 0; i < 2; ++i)
+		{
+			// Just subtract the approximate average of the latest many
+			// audio samples.
+			m_sampAvg[i] = m_sampAvg[i] * (1.f - m_dcCoeff) + tempAdd[i] * m_dcCoeff;
+			tempAdd[i] -= m_sampAvg[i];
+		}
+
+
+		// Increment delay ring buffer index
+		++m_filtFeedbackLoc;
+		if (m_filtFeedbackLoc >= m_delayBufSize)
+		{
+			m_filtFeedbackLoc -= m_delayBufSize;
+		}
+
+		// Send new value to delay line
+		m_filtDelayBuf[0][m_filtFeedbackLoc] = tempAdd[0];
+		m_filtDelayBuf[1][m_filtFeedbackLoc] = tempAdd[1];
+
+		s[0] *= m_outGain;
+		s[1] *= m_outGain;
+
+		// Calculate values for displaying volume in the two volume meters
+		lInPeak = drySignal[0] > lInPeak ? drySignal[0] : lInPeak;
+		rInPeak = drySignal[1] > rInPeak ? drySignal[1] : rInPeak;
+		lOutPeak = s[0] > lOutPeak ? s[0] : lOutPeak;
+		rOutPeak = s[1] > rOutPeak ? s[1] : rOutPeak;
+
+		buf[f][0] = d * buf[f][0] + w * s[0];
+		buf[f][1] = d * buf[f][1] + w * s[1];
+
+		outSum += buf[f][0] + buf[f][1];
+	}
+
+	checkGate(outSum / frames);
+
+	m_phaserControls.m_outPeakL = lOutPeak;
+	m_phaserControls.m_outPeakR = rOutPeak;
+	m_phaserControls.m_inPeakL = lInPeak;
+	m_phaserControls.m_inPeakR = rInPeak;
+
+	return isRunning();
+}
+
+
+
+sample_t PhaserEffect::calcAllpassFilter(sample_t inSamp, sample_rate_t Fs, int filtNum, int channel, float b0, float b1)
+{
+	// The original formula can be found here: https://pastebin.com/AyMH6k36
+	// Much effort was put into CPU optimization, so now the filters
+	// only require a very small number of operations.
+
+	float filterOutput = b0 * (inSamp - m_filt[filtNum][channel].y[1]) +
+		b1 * (m_filt[filtNum][channel].x[0] - m_filt[filtNum][channel].y[0]) +
+		m_filt[filtNum][channel].x[1];
+
+	m_filt[filtNum][channel].x[1] = m_filt[filtNum][channel].x[0];
+	m_filt[filtNum][channel].x[0] = inSamp;
+	m_filt[filtNum][channel].y[1] = m_filt[filtNum][channel].y[0];
+	m_filt[filtNum][channel].y[0] = filterOutput;
+
+	return filterOutput;
+}
+
+
+void PhaserEffect::changeSampleRate()
+{
+	m_lfo->setSampleRate(Engine::mixer()->processingSampleRate());
+	m_twoPiOverSr = F_2PI / Engine::mixer()->processingSampleRate();
+	m_dcCoeff = 44.1f / Engine::mixer()->processingSampleRate();
+	calcAttack();
+	calcRelease();
+
+	for (int i = 0; i < 2; ++i)
+	{
+		m_delayBufSize = Engine::mixer()->processingSampleRate() * 0.03 + 1;
+		m_filtDelayBuf[i].resize(m_delayBufSize);
+	}
+}
+
+
+void PhaserEffect::restartLFO()
+{
+	m_lfo->restart();
+}
+
+
+
+// Takes input of original Hz and the number of cents to detune it by, and returns the detuned result in Hz.
+float PhaserEffect::detuneWithOctaves(float pitchValue, float detuneValue)
```

Comment:
The names are not clear what units ```detuneValue`` is in, the description says cents, the function name says octaves.

Reply:
Ouch, my bad.  I copy+pasted it from other code I had that used cents instead of octaves.